utilities/flink/flink_compaction_filter.cc

```diff
@@ -0,0 +1,148 @@
+  const bool value_or_merge = value_type == ValueType::kValue || value_type == ValueType::kMergeOperand;
+  const bool value_state = state_type == StateType::Value && value_type == ValueType::kValue;
+  const bool list_entry = state_type == StateType::List && value_or_merge;
+  const bool toDecide = value_state || list_entry;
+  const bool list_iter = list_entry && list_element_iter_;
+
+  Decision decision = Decision::kKeep;
+  if (!tooShortValue && toDecide) {
+    decision = list_iter ?
+            ListDecide(existing_value, config, new_value) :
+            Decide(data, config->ttl_, config->timestamp_offset_, CurrentTimestamp(config->useSystemTime_), logger_);
+  }
+  Debug(logger_.get(), "Decision: %d", decision);
+  return decision;
+}
+
+CompactionFilter::Decision FlinkCompactionFilter::ListDecide(
+        const Slice& existing_value, const Config* config, std::string* new_value) const {
+  const int64_t current_timestamp = CurrentTimestamp(config->useSystemTime_);
+  std::size_t offset = 0;
+  if (offset < existing_value.size()) {
+    Decision decision = Decide(existing_value.data(), config->ttl_, offset + config->timestamp_offset_, current_timestamp, logger_);
+    if (decision != Decision::kKeep) {
+      offset = ListNextOffset(existing_value, offset, config->ttl_, current_timestamp);
+      if (offset >= JAVA_MAX_SIZE) {
+        return Decision::kKeep;
+      }
+    }
+  }
+  if (offset >= existing_value.size()) {
+    return Decision::kRemove;
+  } else if (offset > 0) {
+    SetUnexpiredListValue(existing_value, offset, new_value);
+    return Decision::kChangeValue;
+  }
+  return Decision::kKeep;
+}
+
+std::size_t FlinkCompactionFilter::ListNextOffset(
+        const Slice& existing_value, const std::size_t offset, const int64_t ttl, const int64_t current_timestamp) const {
+  std::size_t new_offset = list_element_iter_->NextUnexpiredOffset(existing_value, ttl, current_timestamp);
+  if (new_offset >= JAVA_MAX_SIZE || new_offset < offset) {
+    Error(logger_.get(), "Wrong next offset in list iterator: %d -> %d",
+          offset, new_offset);
+    new_offset = JAVA_MAX_SIZE;
+  } else {
+    Debug(logger_.get(), "Next unexpired offset: %d -> %d",
+          offset, new_offset);
+  }
+  return new_offset;
+}
+
+void FlinkCompactionFilter::SetUnexpiredListValue(
+        const Slice& existing_value, std::size_t offset, std::string* new_value) const {
+  new_value->clear();
+  Slice new_value_slice = Slice(existing_value.data() + offset, existing_value.size() - offset);
+  new_value->assign(new_value_slice.data(), new_value_slice.size());
+  Debug(logger_.get(), "New list value: %s", new_value_slice.ToString(true).c_str());
+}
+
+int64_t FlinkCompactionFilter::CurrentTimestamp(bool useSystemTime) const {
+  using namespace std::chrono;
+  int64_t current_timestamp;
+  if (useSystemTime) {
+    current_timestamp = duration_cast<milliseconds>(system_clock::now().time_since_epoch()).count();
```

Comment:
I would suggest a different approach here. Instead of pushing the TTL time into the filter either on every time change (e.g. for event time) or delegating to system time (procesing time) has some disadvantages. Most importantly it introduced special casing and couples the implementation to the system time for the second case. I would rather suggest that the compaction filter pulls the time once in the beginning of a compation from Flink's TTL time provider. This is a best effort cleanup so that would be ok. Furthermore, the compaction filter can also decide to update the current ttl time from time to time, e.g. after n filter invocations or whatever policy seems good.